lib/cache/collection.go

```diff
@@ -0,0 +1,197 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cache
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+// collection is responsible for managing a resource in the cache.
+type collection[T any, S store[T], U upstream[T]] struct {
+	upstream  U
+	store     S
+	watch     types.WatchKind
+	singleton bool
+}
+
+// upstream is responsible for seeding the cache with resources from
+// the auth server.
+type upstream[T any] interface {
+	getAll(ctx context.Context, loadSecrets bool) ([]T, error)
+}
+
+// store persists the cached resources locally in memory.
+type store[T any] interface {
+	// put will create or update a target resource in the cache.
+	put(t T) error
+	// delete removes a single entry from the cache.
+	delete(t T) error
+	// clear will delete all target resources of the type in the cache.
+	clear() error
+}
+
+type eventHandler interface {
+	// fetch fetches resources and returns a function which will apply said resources to the cache.
+	// fetch *must* not mutate cache state outside of the apply function.
+	// The provided cacheOK flag indicates whether this collection will be included in the cache generation that is
+	// being prepared. If cacheOK is false, fetch shouldn't fetch any resources, but the apply function that it
+	// returns must still delete resources from the backend.
+	fetch(ctx context.Context, cacheOK bool) (apply func(ctx context.Context) error, err error)
+	// onDelete will delete a single target resource from the cache. For
+	// singletons, this is usually an alias to clear.
+	onDelete(t types.Resource) error
+	// onUpdate will update a single target resource from the cache
+	onUpdate(t types.Resource) error
+	// watchKind returns a watch
+	// required for this collection
+	watchKind() types.WatchKind
+}
+
+func (c collection[_, _, _]) watchKind() types.WatchKind {
+	return c.watch
+}
+
+func (c *collection[T, _, _]) onDelete(r types.Resource) error {
+	switch t := r.(type) {
+	case types.Resource153Unwrapper:
+		unwrapped := t.Unwrap()
+		tt, ok := unwrapped.(T)
+		if !ok {
+			return trace.BadParameter("unexpected wrapped type %T (expected %v)", unwrapped, reflect.TypeOf((*T)(nil)).Elem())
+		}
+
+		return trace.Wrap(c.store.delete(tt))
+	case *types.ResourceHeader:
+		return trace.BadParameter("unable to convert types.ResourceHeader to %v", reflect.TypeOf((*T)(nil)).Elem())
+	case T:
+		return trace.Wrap(c.store.delete(t))
+	default:
+		return trace.BadParameter("unexpected type %T (expected %v)", r, reflect.TypeOf((*T)(nil)).Elem())
+	}
+}
```

Comment:
A  lot of delete events only contain `types.ResourceHeader`.  I might be missing something because resource header is clearly being considered here, but won't this logic not work for most resources?